### PR TITLE
[next-auth] follow behavior of js implementation and improve

### PR DESCRIPTION
**Pull request by euxn23**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - https://github.com/nextauthjs/next-auth/blob/main/src/server/index.js
  - https://next-auth.js.org/getting-started/example
  - https://github.com/nextauthjs/next-auth/blob/main/src/providers/index.js
  - https://next-auth.js.org/configuration/providers
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by lluia** [comment]
@euxn23, for the sake of clarification, what's the difference between this PR and #51469 📐 ? For now, I'm gonna review this one, assuming  #51469 is a duplicate 👍🏼

**Comment by balazsorban44** [comment]
sorry for being off-topic, 
@lluia could you please contact me on email or Twitter?

**Comment by euxn23** [comment]
Sorry for my late response.
I fixed merge conflict and rebase master.

@lluia 
#51469 is outdated version, so I had already closed.
I once think I should make this changes splitted into more smaller, but I this 2 commits are my best... 😭 
But I don't think these changes have big risk, because almost of these are only adding more strict type definitions.
If users updates this lib, some type errors may be caused, but I think it's correct breaking... because that code may cause runtime error.

In conclusion, I entrust code owner of DefinitelyTyped, because NextAuth team seems to have no time to consider to adding typings.

**Comment by lluia** [comment]
Thanks for your comment @euxn23; as long as the tests pass, I have the confidence the types shouldn't break in a bad way. 

I'm gonna take the time to review this today with calm; thanks for all the effort you put on it and take time helping bring types up to date 💪🏽

File: types/next-auth/next-auth-tests.ts [hunk 1]
```diff
@@ -5,15 +5,43 @@
...
+    json: Send<T>;
+    status: (statusCode: number) => NextApiResponse<T>;
+    redirect(url: string): NextApiResponse<T>;
+    redirect(status: number, url: string): NextApiResponse<T>;
+    setPreviewData: (data: object | string, options?: {
+        maxAge?: number;
+    }) => NextApiResponse<T>;
+    clearPreviewData: () => NextApiResponse<T>;
+};
+/* ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ */
```
Comment (lluia) [inline types/next-auth/next-auth-tests.ts:43]: We could use here the types from [`./_next.d.ts`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b7fafcf9fc2ea8401fd0ee0e5637e0b73d0f51a1/types/next-auth/_next.d.ts) or is there any specific reason to keep them duplicate? 🧐

File: types/next-auth/providers.d.ts [hunk 2]
```diff
@@ -1,56 +1,86 @@
-import { GenericObject } from './_utils';
+import { User } from '.';
+import { JWT } from './jwt';
+import { NonNullParams, NullableParams, WithAdditionalParams } from './_utils';
 
-export interface Providers {
+export interface Provider<T extends string | undefined = undefined, U = T extends string ? 'oauth' : string> {
+    id: T;
+    name: string;
+    type: U extends string ? U : 'oauth' | 'email' | 'credentials';
```
Comment (lluia) [inline types/next-auth/providers.d.ts:8]: Is there anyway we could simplify the generic here? I got a bit lost trying to follow the logic, I think a small comment explaining the branching would help 📐

**Review by 0ubbe — APPROVED** [review]
I tried to peek at everything and it generally looks very good. 💯 🌟

Thanks for taking the time to go through such an update, I'm a bit worried by the magnitude of the changes, but I see we updated the tests, so we should be cool. 💆🏽‍♂️

I'm happy to approve this and get it published, @euxn23 I would only ask for your availability/help in the next weeks in case issues arise regarding these updates. 

I saw you added yourself as a maintainer of the types 💚 , awesome to have your help. 💪🏽

File: types/next-auth/next-auth-tests.ts (hunk 1 above)
Comment (lluia) [reply types/next-auth/next-auth-tests.ts:43]: @euxn23 noting that I would like to have this answered/addressed before merging 🙏🏽

**Comment by balazsorban44** [comment]
Ready to merge

File: types/next-auth/next-auth-tests.ts (hunk 1 above)
Comment (euxn23) [reply types/next-auth/next-auth-tests.ts:43]: @lluia 

Reason of this is relative import is forbidden by dtslint 😢 

>  Test file should not use a relative import. Use a global import as if this were a user of the package. See: https://github.com/Microsoft/dtslint/blob/master/docs/no-relative-import-in-test.md
